R/pilot_tune_grid.R

```diff
@@ -0,0 +1,182 @@
+  UseMethod("pilot_tune_grid")
+}
+
+#' @export
+pilot_tune_grid.default <- function(object, ...) {
+  msg <- paste0(
+    "The first argument to [pilot_tune_grid()] should be either ",
+    "a model or workflow."
+  )
+}
+
+#' @export
+#' @rdname pilot_tune_grid
+pilot_tune_grid.model_spec <- function(object, preprocessor, resamples, ...,
+                                 param_info = NULL, grid = 10, metrics = NULL,
+                                 control = control_grid(), eval_time = NULL) {
+  if (rlang::is_missing(preprocessor) || !is_preprocessor(preprocessor)) {
+    rlang::abort(paste(
+      "To tune a model spec, you must preprocess",
+      "with a formula or recipe"
+    ))
+  }
+
+  control <- parsnip::condense_control(control, control_grid())
+
+  empty_ellipses(...)
+
+  wflow <- add_model(workflow(), object)
+
+  if (is_recipe(preprocessor)) {
+    wflow <- add_recipe(wflow, preprocessor)
+  } else if (rlang::is_formula(preprocessor)) {
+    wflow <- add_formula(wflow, preprocessor)
+  }
+
+  tune_grid(
+    wflow,
+    resamples = resamples,
+    param_info = param_info,
+    grid = grid,
+    metrics = metrics,
+    control = control,
+    eval_time = eval_time
+  )
+}
+
+#' @export
+#' @rdname pilot_tune_grid
+pilot_tune_grid.workflow <- function(object, resamples, ..., param_info = NULL,
+                               grid = 10, metrics = NULL,
+                               control = control_grid(), eval_time = NULL) {
+  empty_ellipses(...)
+
+  control <- parsnip::condense_control(control, control_grid())
+
+  # Disallow `NULL` grids in `tune_grid()`, as this is the special signal
+  # used when no tuning is required
+  if (is.null(grid)) {
+    rlang::abort(grid_msg)
+  }
+
+  res <-
+    pilot_tune_grid_workflow(
+      object,
+      resamples = resamples,
+      grid = grid,
+      metrics = metrics,
+      pset = param_info,
+      control = control,
+      eval_time = eval_time
+    )
+  .stash_last_result(res)
+  res
+}
+
+# ------------------------------------------------------------------------------
+
+pilot_tune_grid_workflow <- function(workflow,
+                               resamples,
+                               grid = 10,
+                               metrics = NULL,
+                               pset = NULL,
+                               control = control_grid(),
+                               eval_time = NULL,
+                               rng = TRUE) {
+  check_rset(resamples)
+
+
+  metrics <- check_metrics(metrics, workflow)
+  check_eval_time(eval_time, metrics)
+
+  pset <- check_parameters(
+    workflow,
+    pset = pset,
+    data = resamples$splits[[1]]$data,
+    grid_names = names(grid)
+  )
+
+  check_workflow(workflow, pset = pset)
+  check_backend_options(control$backend_options)
+
+  grid <- check_grid(
+    grid = grid,
+    workflow = workflow,
+    pset = pset
+  )
+
+  # Cut grid down to extreme values only for piloting
+  grid <- grid |>
```

Comment:
And here I thought I was being hip and modern in this PR... because I haven't actually converted from base pipe in my own work yet!  I'll swap it back.